sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1798,14 +1837,51 @@ ProgramManager::kernelImplicitLocalArgPos(const std::string &KernelName) const {
   return {};
 }
 
+static bool shouldSkipEmptyImage(sycl_device_binary RawImg) {
+  // For bfloat16 device library image, we should keep it. However, in some
+  // scenario, __sycl_register_lib can be called multiple times and the same
+  // bfloat16 device library image may be handled multiple times which is not
+  // needed. 2 static bool variables are created to record whether native or
+  // fallback bfloat16 device library image has been handled, if yes, we just
+  // need to skip it.
+  sycl_device_binary_property_set ImgPS;
+  static bool IsNativeBF16DeviceLibHandled = false;
+  static bool IsFallbackBF16DeviceLibHandled = false;
+  RTDeviceBinaryImage::PropertyRange DevLibMD(
+      RawImg, __SYCL_PROPERTY_SET_DEVICELIB_METADATA);
+  auto ImgPIt = std::find_if(
+        DevLibMD.begin(), DevLibMD.end(),
+        [](const sycl_device_binary_property Prop) {
+            return Prop->Name && !strcmp("bfloat16", Prop->Name) &&
+                   (Prop->Type == SYCL_PROPERTY_TYPE_UINT32);
+        });
+  if (ImgPIt == DevLibMD.end())
+    return true;
+
+  // A valid bfloat16 device library image is found here, need to check
+  // wheter it has been handled already.
+  uint32_t BF16NativeVal = DeviceBinaryProperty(*ImgPIt).asUint32();
+  if (((BF16NativeVal == 0) && IsFallbackBF16DeviceLibHandled) ||
+      ((BF16NativeVal == 1) && IsNativeBF16DeviceLibHandled))
+    return true;
+
+  if (BF16NativeVal == 0)
+    IsFallbackBF16DeviceLibHandled = true;
+  else
+    IsNativeBF16DeviceLibHandled = true;
+
+  return false;
+  return true;
```

Comment:
```suggestion
```

Reply:
Hi, @steffenlarsen 
The PropertyRange ctor is private, we can't simply call it in any function. I can't come up with any way to bypass this unless we construct RTDeviceBinaryImage based on raw_image but this is what we try to avoid at the beginning. Do you have any good idea?
Thanks very much.

Reply:
Ah, bummer. I suppose the best solution is to revert to what you had before, though I would still keep the use of `DeviceBinaryProperty`.

Reply:
Hi, @steffenlarsen 
Done.
Thanks very much.